File: app/src/main/java/org/schabi/newpipe/info_list/holder/StreamMiniInfoItemHolder.java
package org.schabi.newpipe.info_list.holder;

import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.core.content.ContextCompat;

import org.schabi.newpipe.R;
import org.schabi.newpipe.database.stream.model.StreamStateEntity;
import org.schabi.newpipe.extractor.InfoItem;
import org.schabi.newpipe.extractor.stream.StreamInfoItem;
import org.schabi.newpipe.info_list.InfoItemBuilder;
import org.schabi.newpipe.ktx.ViewUtils;
import org.schabi.newpipe.local.history.HistoryRecordManager;
import org.schabi.newpipe.util.DependentPreferenceHelper;
import org.schabi.newpipe.util.Localization;
import org.schabi.newpipe.util.StreamTypeUtil;
import org.schabi.newpipe.util.image.CoilHelper;
import org.schabi.newpipe.views.AnimatedProgressBar;

import java.util.concurrent.TimeUnit;

public class StreamMiniInfoItemHolder extends InfoItemHolder {
    public final ImageView itemThumbnailView;
    public final TextView itemVideoTitleView;
    public final TextView itemUploaderView;
    public final TextView itemDurationView;
    private final AnimatedProgressBar itemProgressView;

    StreamMiniInfoItemHolder(final InfoItemBuilder infoItemBuilder, final int layoutId,
                             final ViewGroup parent) {
        super(infoItemBuilder, layoutId, parent);

        itemThumbnailView = itemView.findViewById(R.id.itemThumbnailView);
        itemVideoTitleView = itemView.findViewById(R.id.itemVideoTitleView);
        itemUploaderView = itemView.findViewById(R.id.itemUploaderView);
        itemDurationView = itemView.findViewById(R.id.itemDurationView);
        itemProgressView = itemView.findViewById(R.id.itemProgressView);
    }

    public StreamMiniInfoItemHolder(final InfoItemBuilder infoItemBuilder, final ViewGroup parent) {
        this(infoItemBuilder, R.layout.list_stream_mini_item, parent);
    }

    private void updateDurationMarginForProgress() {
        updateDurationMarginForProgress(itemProgressView.getProgress());
    }

    private void updateDurationMarginForProgress(final int progress) {
        if (itemDurationView == null || itemProgressView == null) {
            return;
        }

        final ViewGroup.LayoutParams layoutParams = itemDurationView.getLayoutParams();
        if (!(layoutParams instanceof ViewGroup.MarginLayoutParams)) {
            return;
        }

        final boolean shouldRaiseDuration = itemProgressView.getVisibility() == View.VISIBLE
                && progress > 0;
        final int bottomMargin = itemDurationView.getResources().getDimensionPixelSize(
                shouldRaiseDuration
                        ? R.dimen.stream_thumbnail_duration_margin_with_progress
                        : R.dimen.video_item_search_duration_margin);
        final ViewGroup.MarginLayoutParams marginLayoutParams =
                (ViewGroup.MarginLayoutParams) layoutParams;
        if (marginLayoutParams.bottomMargin != bottomMargin) {
            marginLayoutParams.bottomMargin = bottomMargin;
            itemDurationView.setLayoutParams(marginLayoutParams);
        }
    }

    @Override
    public void updateFromItem(final InfoItem infoItem,
                               final HistoryRecordManager historyRecordManager) {
        if (!(infoItem instanceof StreamInfoItem)) {
            return;
        }
        final StreamInfoItem item = (StreamInfoItem) infoItem;

        itemVideoTitleView.setText(item.getName());
        itemUploaderView.setText(item.getUploaderName());

        if (item.getDuration() > 0) {
            itemDurationView.setText(Localization.getDurationString(item.getDuration()));
            itemDurationView.setBackgroundColor(ContextCompat.getColor(itemBuilder.getContext(),
                    R.color.duration_background_color));
            itemDurationView.setVisibility(View.VISIBLE);

            StreamStateEntity state2 = null;
            if (DependentPreferenceHelper
                    .getPositionsInListsEnabled(itemProgressView.getContext())) {
                state2 = historyRecordManager.loadStreamState(infoItem)
                        .blockingGet()[0];
            }
            if (state2 != null) {
                itemProgressView.setVisibility(View.VISIBLE);
                itemProgressView.setMax((int) item.getDuration());
                itemProgressView.setProgress((int) TimeUnit.MILLISECONDS
                        .toSeconds(state2.getProgressMillis()));
            } else {
                itemProgressView.setVisibility(View.GONE);
            }
        } else if (StreamTypeUtil.isLiveStream(item.getStreamType())) {
            itemDurationView.setText(R.string.duration_live);
            itemDurationView.setBackgroundColor(ContextCompat.getColor(itemBuilder.getContext(),
                    R.color.live_duration_background_color));
            itemDurationView.setVisibility(View.VISIBLE);
            itemProgressView.setVisibility(View.GONE);
        } else {
            itemDurationView.setVisibility(View.GONE);
            itemProgressView.setVisibility(View.GONE);
        }
        updateDurationMarginForProgress();

        // Default thumbnail is shown on error, while loading and if the url is empty
        CoilHelper.INSTANCE.loadThumbnail(itemThumbnailView, item.getThumbnails());

        itemView.setOnClickListener(view -> {
            if (itemBuilder.getOnStreamSelectedListener() != null) {
                itemBuilder.getOnStreamSelectedListener().selected(item);
            }
        });

        switch (item.getStreamType()) {
            case AUDIO_STREAM:
            case VIDEO_STREAM:
            case LIVE_STREAM:
            case AUDIO_LIVE_STREAM:
            case POST_LIVE_STREAM:
            case POST_LIVE_AUDIO_STREAM:
                enableLongClick(item);
                break;
            case NONE:
            default:
                disableLongClick();
                break;
        }
    }

    @Override
    public void updateState(final InfoItem infoItem,
                            final HistoryRecordManager historyRecordManager) {
        final StreamInfoItem item = (StreamInfoItem) infoItem;

        StreamStateEntity state = null;
        if (DependentPreferenceHelper.getPositionsInListsEnabled(itemProgressView.getContext())) {
            state = historyRecordManager
                    .loadStreamState(infoItem)
                    .blockingGet()[0];
        }
        if (state != null && item.getDuration() > 0
                && !StreamTypeUtil.isLiveStream(item.getStreamType())) {
            final int progress = (int) TimeUnit.MILLISECONDS
                    .toSeconds(state.getProgressMillis());
            itemProgressView.setMax((int) item.getDuration());
            if (itemProgressView.getVisibility() == View.VISIBLE) {
                itemProgressView.setProgressAnimated(progress);
            } else {
                itemProgressView.setProgress(progress);
                ViewUtils.animate(itemProgressView, true, 500);
            }
            updateDurationMarginForProgress(progress);
        } else if (itemProgressView.getVisibility() == View.VISIBLE) {
            ViewUtils.animate(itemProgressView, false, 500);
            updateDurationMarginForProgress(0);
        } else {
            updateDurationMarginForProgress(0);
        }
    }

    private void enableLongClick(final StreamInfoItem item) {
        itemView.setLongClickable(true);
        itemView.setOnLongClickListener(view -> {
            if (itemBuilder.getOnStreamSelectedListener() != null) {
                itemBuilder.getOnStreamSelectedListener().held(item);
            }
            return true;
        });
    }

    private void disableLongClick() {
        itemView.setLongClickable(false);
        itemView.setOnLongClickListener(null);
    }
}

File: app/src/main/java/org/schabi/newpipe/info_list/holder/StreamMiniInfoItemHolder.kt
package org.schabi.newpipe.info_list.holder

import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import org.schabi.newpipe.R
import org.schabi.newpipe.database.stream.model.StreamStateEntity
import org.schabi.newpipe.extractor.InfoItem
import org.schabi.newpipe.extractor.stream.StreamInfoItem
import org.schabi.newpipe.extractor.stream.StreamType
import org.schabi.newpipe.info_list.InfoItemBuilder
import org.schabi.newpipe.ktx.animate
import org.schabi.newpipe.local.history.HistoryRecordManager
import org.schabi.newpipe.util.DependentPreferenceHelper
import org.schabi.newpipe.util.Localization
import org.schabi.newpipe.util.StreamTypeUtil
import org.schabi.newpipe.util.image.PicassoHelper
import org.schabi.newpipe.views.AnimatedProgressBar
import java.util.concurrent.TimeUnit

open class StreamMiniInfoItemHolder internal constructor(infoItemBuilder: InfoItemBuilder, layoutId: Int,
                                                         parent: ViewGroup?) : InfoItemHolder(infoItemBuilder, layoutId, parent) {
    val itemThumbnailView: ImageView
    val itemVideoTitleView: TextView
    val itemUploaderView: TextView
    val itemDurationView: TextView
    private val itemProgressView: AnimatedProgressBar

    init {
        itemThumbnailView = itemView.findViewById(R.id.itemThumbnailView)
        itemVideoTitleView = itemView.findViewById(R.id.itemVideoTitleView)
        itemUploaderView = itemView.findViewById(R.id.itemUploaderView)
        itemDurationView = itemView.findViewById(R.id.itemDurationView)
        itemProgressView = itemView.findViewById(R.id.itemProgressView)
    }

    constructor(infoItemBuilder: InfoItemBuilder, parent: ViewGroup?) : this(infoItemBuilder, R.layout.list_stream_mini_item, parent)

    public override fun updateFromItem(infoItem: InfoItem?,
                                       historyRecordManager: HistoryRecordManager) {
        if (!(infoItem is StreamInfoItem)) {
            return
        }
        val item: StreamInfoItem = infoItem
        itemVideoTitleView.setText(item.getName())
        itemUploaderView.setText(item.getUploaderName())
        if (item.getDuration() > 0) {
            itemDurationView.setText(Localization.getDurationString(item.getDuration()))
            itemDurationView.setBackgroundColor(ContextCompat.getColor(itemBuilder.getContext(),
                    R.color.duration_background_color))
            itemDurationView.setVisibility(View.VISIBLE)
            var state2: StreamStateEntity? = null
            if (DependentPreferenceHelper.getPositionsInListsEnabled(itemProgressView.getContext())) {
                state2 = historyRecordManager.loadStreamState(infoItem)
                        .blockingGet().get(0)
            }
            if (state2 != null) {
                itemProgressView.setVisibility(View.VISIBLE)
                itemProgressView.setMax(item.getDuration().toInt())
                itemProgressView.setProgress(TimeUnit.MILLISECONDS
                        .toSeconds(state2.getProgressMillis()).toInt())
            } else {
                itemProgressView.setVisibility(View.GONE)
            }
        } else if (StreamTypeUtil.isLiveStream(item.getStreamType())) {
            itemDurationView.setText(R.string.duration_live)
            itemDurationView.setBackgroundColor(ContextCompat.getColor(itemBuilder.getContext(),
                    R.color.live_duration_background_color))
            itemDurationView.setVisibility(View.VISIBLE)
            itemProgressView.setVisibility(View.GONE)
        } else {
            itemDurationView.setVisibility(View.GONE)
            itemProgressView.setVisibility(View.GONE)
        }

        // Default thumbnail is shown on error, while loading and if the url is empty
        PicassoHelper.loadThumbnail(item.getThumbnails()).into(itemThumbnailView)
        itemView.setOnClickListener(View.OnClickListener({ view: View? ->
            if (itemBuilder.getOnStreamSelectedListener() != null) {
                itemBuilder.getOnStreamSelectedListener()!!.selected(item)
            }
        }))
        when (item.getStreamType()) {
            StreamType.AUDIO_STREAM, StreamType.VIDEO_STREAM, StreamType.LIVE_STREAM, StreamType.AUDIO_LIVE_STREAM, StreamType.POST_LIVE_STREAM, StreamType.POST_LIVE_AUDIO_STREAM -> enableLongClick(item)
            StreamType.NONE -> disableLongClick()
            else -> disableLongClick()
        }
    }

    public override fun updateState(infoItem: InfoItem,
                                    historyRecordManager: HistoryRecordManager) {
        val item: StreamInfoItem = infoItem as StreamInfoItem
        var state: StreamStateEntity? = null
        if (DependentPreferenceHelper.getPositionsInListsEnabled(itemProgressView.getContext())) {
            state = historyRecordManager
                    .loadStreamState(infoItem)
                    .blockingGet().get(0)
        }
        if ((state != null) && (item.getDuration() > 0
                        ) && !StreamTypeUtil.isLiveStream(item.getStreamType())) {
            itemProgressView.setMax(item.getDuration().toInt())
            if (itemProgressView.getVisibility() == View.VISIBLE) {
                itemProgressView.setProgressAnimated(TimeUnit.MILLISECONDS
                        .toSeconds(state.getProgressMillis()).toInt())
            } else {
                itemProgressView.setProgress(TimeUnit.MILLISECONDS
                        .toSeconds(state.getProgressMillis()).toInt())
                itemProgressView.animate(true, 500)
            }
        } else if (itemProgressView.getVisibility() == View.VISIBLE) {
            itemProgressView.animate(false, 500)
        }
    }

    private fun enableLongClick(item: StreamInfoItem) {
        itemView.setLongClickable(true)
        itemView.setOnLongClickListener(OnLongClickListener({ view: View? ->
            if (itemBuilder.getOnStreamSelectedListener() != null) {
                itemBuilder.getOnStreamSelectedListener()!!.held(item)
            }
            true
        }))
    }

    private fun disableLongClick() {
        itemView.setLongClickable(false)
        itemView.setOnLongClickListener(null)
    }
}
